### PR TITLE
Redo some build system stuff

### PR DESCRIPTION
**Pull request by bennybp**
# Description of Pull Request

Redo how the core pulsar library was built (pulsar itself, not the superbuild).

The subdirectories are no longer built as independent SO files. Instead, all the files are pulled into the top-level cmake and built there, resulting in a single pulsar_core.so file. Python submodules are handled with pybind11's def_submodule

Benefits:
  * Ensures flags, includes, libraries, etc, are consistent across the entire project
  * Prevents some complexity with some core modules requiring an external project that the others don't (particularly if there are RPATH issues)
  * Entire build process is basically in a single cmake file (although it could be split up a little)
  * No changes required to modules/StandardModules, etc

Cons:
  * A bit less flexible. However, that should be ok for the core (ie, might be hard to add another language or to have some file-specific compilation options).
  * Python import stuff is a little weird (for now)

Any comments/questions from @loriab  or @ryanmrichard ?

### CODE REVIEW

**Comment by ryanmrichard** [comment]
The CMake thing looks good to me (and honestly makes a lot more sense now that you have brought it up).  Another con for the list is that you have to re-link the entire project now instead of just say the system module (probably a moot point because linking is nearly instantaneous).

Can you elaborate a bit more on the python change?  I looked at the code and it's not clear to me what the difference is?  Or is there no difference aside from Pybind 11 now offers this sub_module function?

File: pulsar/CMakeLists.txt [hunk 1]
```diff
@@ -66,9 +61,165 @@ add_subdirectory(system)
...
+#################
+# Includes
+#################
+# We need to add the parent of this directory
+get_filename_component(PULSAR_CORE_PARENT_DIR ${CMAKE_SOURCE_DIR} DIRECTORY)
+list(APPEND PULSAR_CORE_CXX_INCLUDES ${PULSAR_CORE_PARENT_DIR})
+
+# External dependencies (mark as system includes)
+list(APPEND PULSAR_CORE_CXX_EXTERNAL_INCLUDES ${MEMWATCH_INCLUDE_DIR})
+list(APPEND PULSAR_CORE_CXX_EXTERNAL_INCLUDES ${Boost_INCLUDE_DIRS})
```
Comment (loriab) [inline pulsar/CMakeLists.txt]: I thought Boost had been booted.

Comment (bennybp) [inline pulsar/CMakeLists.txt]: I think we still have some boost::graph stuff in there?

**Comment by bennybp** [comment]
The python difference is that everything is loaded by the top-level `__init__` file, but still remains under the pulsar_core module/namespace. Then it is selectively exported within the other `__init__` files.

For example, `pulsar.pulsar_core.datastore` is exported to `pulsar.datastore` via `datastore/__init__.py`, which needs to to a relative import from its parent.

File: pulsar/CMakeLists.txt (hunk 1 above)
Comment (loriab) [inline pulsar/CMakeLists.txt]: Ok, I guess it was compiled boost libraries that I was remembering. Yes, everything looks fine by me. `__init__` business is perhaps a little funny, but I don't think it's flaky.

**Comment by ryanmrichard** [comment]
Ok, I think I quasi understand.

**Pull request merged**